sklearn/cluster/k_means_.py

```diff
@@ -391,9 +393,7 @@ def _kmeans_single_elkan(X, n_clusters, max_iter=300, init='k-means++',
                          verbose=False, x_squared_norms=None,
                          random_state=None, tol=1e-4,
                          precompute_distances=True):
-    if sp.issparse(X):
-        raise ValueError("algorithm='elkan' not supported for sparse input X")
```

Comment:
Sorry for confusing things, but I think this error message is better, being more specific and helpful to the user. So please keep this and, as you suggested, remove the check_array